master/pkg/schemas/expconf/log_pattern_config.go

```diff
@@ -3,87 +3,262 @@ package expconf
 import (
 	"encoding/json"
 	"fmt"
+	"sort"
 
 	log "github.com/sirupsen/logrus"
 
+	"github.com/pkg/errors"
+
+	"github.com/determined-ai/determined/master/pkg/ptrs"
 	"github.com/determined-ai/determined/master/pkg/schemas"
-	"github.com/determined-ai/determined/master/pkg/union"
+	"github.com/determined-ai/determined/master/pkg/set"
 )
 
 // LogPoliciesConfigV0 is a list of log policies.
+//
+//go:generate ../gen.sh
 type LogPoliciesConfigV0 []LogPolicyV0
 
-// Merge implemenets the mergable interface.
+// WithDefaults implements the Defaultable pseudo-interface.
+func (b LogPoliciesConfigV0) WithDefaults() LogPoliciesConfigV0 {
+	defaultPolicies := LogPoliciesConfigV0{
+		LogPolicyV0{
+			RawName:    ptrs.Ptr(CUDAOOM),
+			RawPattern: ptrs.Ptr(CUDAOOMPattern),
+		},
+		LogPolicyV0{
+			RawName:    ptrs.Ptr(ECCError),
+			RawPattern: ptrs.Ptr(ECCErrorPattern),
+		},
+	}
+
+	return schemas.Merge(b, defaultPolicies)
+}
+
+// Merge implements the Mergable pseudo-interface.
+// Union merge log policies.
+// Special cases:
+// 1. We may see policies with different names same patterns. We keep both of them, but which name
+// will be shown in the UI is undefined.
+// 2. There could be policies with the same name different patterns. We save the one with the higher priority.
+// Unsetting default values depends on this behavior.
+// 2.1 Polices that don't have name but have same patterns is a special case. These are legacy
+// policies.
 func (b LogPoliciesConfigV0) Merge(
```

Comment:
This seems more complicated that I think it ought to be.

I expected something like this:

```go
names := map[string]bool{}
unnamed := map[string]bool{}

stringify := func(s string) string { return

// keep everything in b unconditionally
out := append(LogPoliciesConfigV0{}, b)
for _, lp := range b {
  if lp.RawName == nil {
    unnamed[fmt.Sprintf("%v:%v", lp.RawAction, lp.RawPattern)] = true
  } else {
    names[*lp.RawName] = true
  }
}

// add anything in src that is new
for _, lp := range src {
  if lp.RawName == nil {
    s := fmt.Sprintf("%v:%v", lp.RawAction, lp.RawPattern)
    if _, ok := unnamed[s]; !ok {
      out = append(out, lp)
      unnamed[s] = true
    }
  } else if _, ok := names[*lp.RawName]; !ok {
    out = append(out, lp)
    names[*lp.RawName] = true
  }
}

return out
```

And now that I think about it, if you do it this way... do you ever need to sort?  This is like a one-pass source-order-preserving merge.

Reply:
Neat!